packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,349 @@
+   * When passing an array, the first supported locale is used.
+   * Defaults to `navigator.language` in supported environments.
+   */
+  locale?: Locale;
+  /**
+   * Prevent users from closing the modal by clicking/tapping the overlay or
+   * pressing the escape key. Default `false`.
+   */
+  preventClose?: boolean;
+  /**
+   * Enables focusing a particular element in the dialog content and override default behavior
+   */
+  initialFocusRef?: RefObject<HTMLElement>;
+  /**
+   * a ReactNode or a function that returns the content of the modal dialog.
+   */
+  children?:
+    | ReactNode
+    | (({ onClose }: { onClose?: DialogProps['onClose'] }) => ReactNode);
+  /**
+   * The duration of the dialog animation in milliseconds.
+   * If you wish to animate the dialog, provide a value of the animation duration to enable the animation to complete before the dialog closes.
+   * default: 0
+   */
+  animationDuration?: number;
+  [key: DataAttribute]: string | undefined;
+}
+
+export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
+  (props, ref) => {
+    const {
+      open,
+      isModal = false,
+      children,
+      onClose,
+      closeButtonLabel,
+      className,
+      initialFocusRef,
+      preventClose,
+      animationDuration = 0,
+      onCloseStart,
+      ...rest
+    } = props;
+    const dialogRef = useRef<HTMLDialogElement>(null);
+    const openRef = useLatest<boolean>(open);
+    const lastFocusedElementRef = useRef<HTMLElement | null>(null);
+
+    // eslint-disable-next-line compat/compat
+    const hasNativeDialog = window.HTMLDialogElement !== undefined;
+
+    useScrollLock(open && isModal);
+
+    // set initial focus on the modal dialog content
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+      let timeoutId: NodeJS.Timeout;
+      if (open && dialogElement) {
+        timeoutId = setTimeout(() => {
+          if (initialFocusRef?.current) {
+            initialFocusRef?.current?.focus();
+          } else {
+            const firstFocusableElement = getFirstFocusableElement(
+              dialogElement,
+              isModal,
+            );
+            if (firstFocusableElement) {
+              firstFocusableElement.focus();
+            } else {
+              dialogElement.focus();
+            }
+          }
+        }, animationDuration);
+      }
+      return () => {
+        clearTimeout(timeoutId);
+      };
+    }, [open, initialFocusRef?.current, isModal, animationDuration]);
+
+    useEffect(() => {
+      // restore focus to opening element
+      if (open) {
+        if (document.activeElement instanceof HTMLElement) {
+          lastFocusedElementRef.current = document.activeElement;
+        }
+      }
+      return () => {
+        if (lastFocusedElementRef.current) {
+          setTimeout(
+            () => lastFocusedElementRef.current?.focus(),
+            animationDuration,
+          );
+        }
```

Comment:
This should only run if `open` is `false`. It could be `true` if only the `animationDuration` has changed.

Reply:
Although then it might not run if the Dialog component is unmounted without setting `open` to `false` first 🤔  

I'd suggest wrapping `animationDuration` in `useLatest` so that you can pass it to the hook without having to re-run the hook when it changes.